Backports 0.17.1

haampie opened the pull request:
PR with backports cherrypicked from the develop branch for the 0.17.1 release

Included PRs:
- [x] #25764
- [x] #27846 
- [x] #24675
- [x] #25961
- [x] #27255
- [x] #27472
- [x] #27379
- [x] #27408
- [x] #27676
- [x] #21447
- [x] #27713
- [x] #27707
- [x] #27911
- [x] #26996 
- [x] #27970
- [x] #28031
- [x] #27837 
- [x] #21682
- [x] #27625
- [x] #28037
- [x] #28036
- [x] #14686
- [x] #28072
- [x] #28113

dylex committed:
make --enable-locks actually enable locks (#24675)

robertu94 committed:
openmpi fix external find for 0.17 (#27255)

Jordan474 committed:
Fix log-format reporter ignoring install errors (#25961)

When running `spack install --log-format junit|cdash ...`, install
errors were ignored. This made spack continue building dependents of
failed install, ignoring `--fail-fast`, and exit 0 at the end.

bonachea committed:
upcxx: Update the UPC++ package to 2021.9.0 (#26996)

* upcxx: Update the UPC++ package to 2021.9.0

* Add the new release, and a missing older one.

* Remove the spack package cruft for supporting the obsolete build system that
  was present in older versions that are no longer supported.

* General cleanups.

Support for library versions older than 2020.3.0 is officially retired,
for two reasons:

1. Releases prior to 2020.3.0 had a required dependency on Python 2,
   which is [officially EOL](https://www.python.org/doc/sunset-python-2/)
   as of Jan 1 2020, and is no longer considered secure.
2. (Most importantly) The UPC++ development team is unable/unwilling to
   support releases more than two years old.  UPC++ provides robust
   backwards-compatibility to earlier releases of UPC++ v1.0, with very
   rare well-documented/well-motivated exceptions.  Users are strongly
   encouraged to update to a current version of UPC++.

NOTE: Most of the lines changed in this commit are simply re-indentation,
and thus might be best reviewed in a diff that ignores whitespace.

* upcxx: Detect Cray XC more explicitly

This change is necessary to prevent false matches occuring on new Cray Shasta
systems, which do not use the aries network but were incorrectly being treated
as a Cray XC + aries platform.

UPC++ has not yet deployed official native support for Cray Shasta, but this
change is sufficient to allow building the portable backends there.

haampie committed:
Fix overloaded argparse keys (#27379)

Commands should not reuse option names defined in main.

haampie committed:
Turn some verbose messages into debug messages (#27408)

alalazo commented on .github/workflows/unit_tests.yaml:
```diff
@@ -199,6 +199,11 @@ jobs:
     runs-on: ubuntu-latest
     container: spack/github-actions:centos6
     steps:
+    - name: Setup git configuration
+      run: |
+        # Need this for the git tests to succeed.
+        git --version
+        . .github/workflows/setup_git.sh
```
Here we are trying to source a file before cloning the Spack repository. I think this line needs to be added twice in each branch of the test (with and without coverage). 

It's unfortunate, but `centos6` is so old that we can't use the `actions/checkout` action inside it if I recall correctly, so we resort to cloning Spack manually.

haampie committed:
ci: run style unit tests only if we target develop (#27472)

Some tests assume the base branch is develop, but this branch may not
have been checked out.

haampie committed:
Use bash in setup_git.sh (#27676)

Paul-Ferrell committed:
Handle byte sequences which are not encoded as UTF8 while logging. (#21447)

Fix builds which produce a lines with non-UTF8 output while logging
The alternative is to read in binary mode, and then decode while
ignoring errors.

alalazo committed:
spack audit: fix API calls (#27713)

This broke in #24858

becker33 committed:
Improve debug info from concretizer (#27707)

haampie committed:
Install dir creation message demoted to "debug" level (#27911)

becker33 committed:
Add option to minimize full debug cores. include warning message about performance (#27970)

Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>

haampie committed:
Revert "patches: make re-applied patches idempotent (#26784)" (#27625)

This reverts commit c5ca0db27fce5d772dc8a4fcffec3b62bb0bf1f3.

tldahlgren committed:
Provide meaningful message for empty environment installs (#28031)

* Provide a meaningful failure message for installation of an empty environment

* Allow regenerating view per offline discussion

mslacken committed:
Added opensuse/leap:15  to spack containerize (#27837)

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

aweits committed:
MANPATH needs a trailing ':' to utilize system defaults (#21682)

otherwise spack breaks using system man pages by default.

Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>

haampie committed:
Fix table formatting (#28037)

haampie committed:
Set backup=False by default in filter_file (#28036)

victorusu committed:
Add setdefault option to tcl module (#14686)

This commit introduces the command

spack module tcl setdefault <package>

similar to the one already available for lmod

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

netromdk committed:
py-vermin: add latest version 1.3.2 (#28072)

* py-vermin: add latest version 1.3.2

* [vermin] Fixed usages of super() without arguments (v3+)

spenpw committed:
sbang: respect package permissive package permissions for sbang (#25764)

Co-authored-by: Todd Gamblin <tgamblin@llnl.gov>
Co-authored-by: Tamara Dahlgren <35777542+tldahlgren@users.noreply.github.com>

alalazo committed:
py-pandas: fix issue reported by vermin

alalazo commented:
@bonachea We are releasing a patch version of Spack at the end of the week. I have backported 159f4c13f6e6bc57f59bc085cd29a8ce7bea7694 for UPC++. Is that fine with you or are there any more updates needed for the package?

bonachea commented:
> @bonachea We are releasing a patch version of Spack at the end of the week. I have backported [159f4c1](https://github.com/spack/spack/commit/159f4c13f6e6bc57f59bc085cd29a8ce7bea7694) for UPC++. Is that fine with you or are there any more updates needed for the package?

@alalazo Looks good to me - thanks! 
We don't have any other changes immediately pending.

haampie committed:
Regenerate views when specs already installed (#28113)

With this commit:
```
$ spack env activate --temp
$ spack install zlib
==> All of the packages are already installed
==> Updating view at /tmp/spack-faiirgmt/.spack-env/view
$ spack install zlib
==> All of the packages are already installed
```

Before this PR:
```
$ spack env activate --temp
$ spack install zlib
==> All of the packages are already installed
$ spack install zlib
==> All of the packages are already installed
```

No view was generated

trws committed:
locks: allow locks to work under high contention (#27846)

* locks: allow locks to work under high contention

This is a bug found by Harshitha Menon.  

The `lock=None` line shouldn't be a release but should be 
```
return (lock_type, None)
``` 
to inform the caller it couldn't get the lock type requested without
disturbing the existing lock object in the database.  There were also a
couple of bugs due to taking write locks at the beginning without any
checking or release, and not releasing read locks before requeueing.
This version no longer gives me read upgrade to write errors, even
running 200 instances on one box.

* Change lock in check_deps_status to read, release if not installed,
  not sure why this was ever write, but read definitely is more
  appropriate here, and the read lock is only held out of the scope if
  the package is installed.

* Release read lock before requeueing to reduce chance of livelock, the
  timeout that caused the original issue now happens in roughly 3 of 200
  workers instead of 199 on average.

alalazo commented:
@spackbot run pipeline

alalazo committed:
Bump version and update CHANGELOG.md
